registry/client_commands_test.go

```diff
@@ -66,95 +61,252 @@ func registryMockup(ctx context.Context, t *testing.T, testName string) *httptes
+// 	ns, err = getNamespaceByPrefix(prefix)
+// 	require.NoError(t, err)
+
+// 	expectedJwks := jwk.NewSet()
+// 	pubKey3, err := jwk.PublicKeyOf(privKey3)
+// 	require.NoError(t, err)
+// 	err = expectedJwks.AddKey(pubKey3)
+// 	require.NoError(t, err)
+// 	err = expectedJwks.AddKey(pubKey4)
+// 	require.NoError(t, err)
+// 	expectedJwksBytes, err := json.Marshal(expectedJwks)
+// 	require.NoError(t, err)
+// 	expectedJwksStr := string(expectedJwksBytes)
+
+// 	expectedKids, err := getSortedKids(ctx, expectedJwksStr)
+// 	require.NoError(t, err)
+// 	actualKids, err := getSortedKids(ctx, ns.Pubkey)
+// 	require.NoError(t, err)
+// 	require.Equal(t, expectedKids, actualKids)
+// }
 
-	_, err := config.GetIssuerPublicJWKS()
-	require.NoError(t, err)
-	privKey, err := config.GetIssuerPrivateJWK()
-	require.NoError(t, err)
-
-	//Test functionality of registering a namespace (without identity)
-	err = NamespaceRegister(privKey, svr.URL+"/api/v1.0/registry", "", "/foo/bar", "mock_site_name")
-	require.NoError(t, err)
-
-	//Test we can list the namespace without an error
-	t.Run("Test namespace list", func(t *testing.T) {
-		//Set up a buffer to capture stdout
-		var stdoutCapture string
-		oldStdout := os.Stdout
-		r, w, _ := os.Pipe()
-		os.Stdout = w
-
-		//List the namespaces
-		err = NamespaceList(svr.URL + "/api/v1.0/registry")
-		require.NoError(t, err)
-		w.Close()
-		os.Stdout = oldStdout
-
-		capturedOutput := make([]byte, 1024)
-		n, _ := r.Read(capturedOutput)
-		stdoutCapture = string(capturedOutput[:n])
-		assert.Contains(t, stdoutCapture, `"prefix":"/foo/bar"`)
-	})
-
-	t.Run("Test register namespace with sitename", func(t *testing.T) {
-		res, err := http.Get(svr.URL + "/api/v1.0/registry/foo/bar")
-		require.NoError(t, err)
-		require.Equal(t, http.StatusOK, res.StatusCode)
-		data, err := io.ReadAll(res.Body)
-		require.NoError(t, err)
-		ns := server_structs.Namespace{}
-		err = json.Unmarshal(data, &ns)
-		require.NoError(t, err)
-		assert.Equal(t, ns.AdminMetadata.SiteName, "mock_site_name")
-	})
-
-	t.Run("Test namespace delete", func(t *testing.T) {
-		//Test functionality of namespace delete
-		err = NamespaceDelete(svr.URL+"/api/v1.0/registry/foo/bar", "/foo/bar")
-		require.NoError(t, err)
-		var stdoutCapture string
-		oldStdout := os.Stdout
-		r, w, _ := os.Pipe()
-		os.Stdout = w
-		err = NamespaceGet(svr.URL + "/api/v1.0/registry")
-		require.NoError(t, err)
-		w.Close()
-		os.Stdout = oldStdout
-
-		capturedOutput := make([]byte, 1024)
-		n, _ := r.Read(capturedOutput)
-		stdoutCapture = string(capturedOutput[:n])
-		assert.Equal(t, "[]\n", stdoutCapture)
-	})
+func TestRegistryKeyChainingOSDF(t *testing.T) {
 	server_utils.ResetTestState()
-}
 
-func TestRegistryKeyChainingOSDF(t *testing.T) {
 	ctx, cancel, egrp := test_utils.TestContext(context.Background(), t)
-	defer func() { require.NoError(t, egrp.Wait()) }()
-	defer cancel()
+	t.Cleanup(func() {
+		func() { assert.NoError(t, egrp.Wait()) }()
+		cancel()
```

Comment:
This is clearly the deadlock and looks like some bad copy/pasting without understanding what the original code does.

The original was:
```
	defer func() { require.NoError(t, egrp.Wait()) }()
	defer cancel()
```

Since `defer` is executed LIFO, this will invoke `cancel()` followed by the `Wait()` (to wait on the canceled functions to clean up).

The revised code is effectively this:

```
         defer func() {
	        func() { require.NoError(t, egrp.Wait()) }()
	        cancel()
          }
```

Which *waits* on the group to stop, then shuts it down, a clear deadlock.

Reply:
Thanks for pointing out and fixing this lingering problem! Having said that, this is just an intermediate problem in the tests - the root problem is topology mock server doesn't response in `TestRegistryKeyChainingOSDF` and test timed out in `TestRegistryKeyChaining` (see the error logs in the [tests result](https://github.com/PelicanPlatform/pelican/actions/runs/12469465642/job/34804139018)). These two errors were suppressed if the other two tests in client_commands_test.go are commented out. Justin figured out there is a test state leak because commenting out the other tests allowed both TestRegistryKeyChaining and TestRegistryKeyChainingOSDF to pass. I'm still looking for the leak and hoping fixing the problems mentioned in your new PR comments could potentially fix the leak.

Reply:
<google-sheets-html-origin><style type="text/css"><!--td {border: 1px solid #cccccc;}br {mso-data-placement:same-cell;}--></style>
  | Test result | Logs
-- | -- | --
Comment Out ⬇️ | TestRegistryKeyChainingOSDF | TestRegistryKeyChaining |  
TestServeNamespaceRegistry | ❌ | ❌ | https://github.com/PelicanPlatform/pelican/actions/runs/12471626378/job/34808934516
TestMultiPubKeysRegisteredOnNamespace | ❌ | ❌ | https://github.com/PelicanPlatform/pelican/actions/runs/12471246608/job/34807853287
Both | ✔️ | ✔️ | https://github.com/PelicanPlatform/pelican/actions/runs/12447058854/job/34750148518